_posts/2022-08-29-sbt-tutorial.md

```diff
@@ -0,0 +1,514 @@
+```
+lazy val commonSettings = Seq(
+  target := { baseDirectory.value / "target2" }
+)
+lazy val module_1 = (project in file("module-1"))
+  .settings(
+    commonSettings,
+    // other settings
+  )
+```
+
+- IN the similar way, extract the common dependency libraries into a single place and re-use. 
+```
+val commonLibs = Seq(
+  "com.typesafe" % "config" % "1.4.2"
+)
+
+lazy val module_2 = (project in file("module-2"))
+.settings(
+  libraryDependencies ++= commonLibs ++ Seq("com.lihaoyi" %% "fansi" % "0.4.0")
+)
+```
+
+## Executing commands on each Module
+Now that we are ready with a multi module project, let's see how we can execute sbt commands module wise. 
+On the root of the project, if we execute _sbt_ it will start the sbt session. When we run the compile command, it will compile all the modules.
+
+Once we are inside sbt session, we can switch to a particular sub-module using:
+```
+project module_2
+```
+Now, when we compile, only this module will get compiled. But, if this module depends on another module, sbt will compile that module as well.
+
+## Plugins
+One of the most important features of SBT is its support for plugins. Plugins helps to extend the sbt with custom features which can be published and shared between multiple teams. 
+For handling plugins, a special file called as _plugins.sbt_ is used. This file is kept under that directory _project_. Some of the common usages of plugins are:
+ - packaging plugins to create jar, exe and other types
+ - static code analysis plugins
+ - code generation plugins
+
+ Let's look at an example of _sbt-assembly_ plugin. This plugin helps to create an executable jar file from the sbt project. 
+
+ As a first step, we need to create _plugins.sbt_ and add the plugin definition to it:
+ ```
+ addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "1.2.0")
+ ```
+ As next step, we need to enable the required configurations needed for this plugin to work. To run a jar, we need to provide the main class to use. We can do that by providing the mainClass in the settings in build.sbt for required module(sometimes, we need to create jar for a sub module, sometime for the  combined project)
+
+ ```
+assembly / mainClass := Some("com.rockthejvm.Module2Main"),
+ ```
+
+ Now we can use the sbt command `assembly`. This will create the jar file under the relevabt project's target folder. 
+ Similar to `assembly/mainClass` there are many other configurations to customise the jar creation.
+
+ This is just one of the plugins. There are many plugins available which can help to improve the development experience.
+
+ ## Global plugins
+ In the previous section, we added the plugin to the project. Sometimes, we might need to add some plugins irrespective of the project. For example, there may be some plugin to publish the apps to an internal repository. This need not be kept in the git repo, instead can be shared across all the repositories. 
```

Comment:
- the plugin -> a plugin
- irrespective of the project -> that have nothing to do with the project itself